contribute/design/1.x/cstor/pool/pool_expansion_by_expanding_disk.md

```diff
@@ -0,0 +1,135 @@
+
+Below is high-level info to explain the Day2 operations that are currently supported.
+
+- Pool Expansion When BlockDevice/raid group added:
+	cstor-pool-mgmt will iterate over all the BlockDevices that exist in CStorPoolInstance resource and verify whether the corresponding BlockDevice/raid group is part of the pool. If that BlockDevice/raid group is not part of the pool then pool expansion will be triggered on newly added BlockDevice/raid group.
+
+- BlockDevice Replacement In Pool:
+    cstor-pool manager will iterate over all the BlockDevices present in CstorPoolInstance and verifies whether the current blockdevice is for BlockDevice replacement request[How cstor-pool-mgmt can know? We are maintaining annotation called openebs.io/predecessor: <replaced_blockdevice_name> on the claim of that blockdevice]? If requested BlockDevice is valid for replacement then the pool manager will perform BlockDevice replacement operation on the pool.
+
+### Proposed Implementations
+
+- CStor-pool Manger inbuilt CSPI controller watches for CSPI resources and tries to get the desired state via every sync operation. As part of reconciliation logic, it will iterate over all the BlockDevices that exist on CStorPoolInstance resource if it found BlockDevice resource capacity(considerable changes) is greater than the existing capacity of BlockDevice on CstorPoolInstanceBlockDevice then we will follow below-mentioned steps to perform the [pool expansion](#Pool_Expansion_Steps).
+- [What if cStor is consuming part of partitioned disk/entire partitioned disk? The User has to extend the partition manually after that NDM will update the capacity on the corresponding blockdevice CR. Once NDM updates the capacity on BlockDevice in next reconciliation of CSPI Pool Expansion will be triggered].
+- [What if only one blockdevice got expanded in case of raid/mirror configuration? Pool expansion operation will be triggered but there won't be any change in the size of the pool].
+
+Currently CstorPoolInstanceBlockDevice holds below configurations 
+```go
+// CStorPoolInstanceBlockDevice contains the details of block devices that
+// constitutes a raid group.
+type CStorPoolInstanceBlockDevice struct {
+	// BlockDeviceName is the name of the block device.
+	BlockDeviceName string `json:"blockDeviceName"`
+
+	// Capacity is the capacity of the block device.
+	// It is system generated
+	Capacity string `json:"capacity"`
+
+	// DevLink is the dev link for block devices
+	DevLink string `json:"devLink"`
+}
+```
+
+### Steps to perform user stories
+
+- User has to make sure that blockdevice capacity to needs to be updated.
+
+#### Pool Expansion Steps:
+
+Step1: Remove the buffering partition if exists which was created by cstor-pool container(ex: sdb9).
```

Comment:
Does this mean that, even if cstor is using a complete disk, to perform expansion, we need to remove one partition created by cstor and expand the other? Doesn't zfs automatically takes care of that?

Reply:
No zfs will not do it automatically!! Because of some loopholes(there might be cases where zfs using partitioned disk if autoexpand is set and if triggering `zpool online expansion` expands the pool then it will over right other partitioned disks which might be using for some other purpose).

Reply:
One more question @mittachaitu .
> Does this mean that, even if cstor is using a complete disk, to perform expansion, we need to remove one partition created by cstor and expand the other?

Is this the step do expansion.?

Reply:
Yes!!